smdebug/tensorflow/keras.py

```diff
@@ -711,6 +714,104 @@ def _remove_fetches_and_callbacks(self, mode):
             x.fetch_callbacks.pop(tf_obj)
         self._fetches_added.clear()
 
+    def _decrement_step(self):
+        # Called when both profiler and debugger are enabled in the native training loop
+        # to adjust the step number
+        self.step -= 1
+        self.mode_steps[self.mode] -= 1
+
+        # Increment Global step number irrespective of what mode it is
+        if self.mode != ModeKeys.GLOBAL:
+            self.mode_steps[ModeKeys.GLOBAL] = self.step
+
+    def _start_phase_python_profiling(self, mode):
+        if self.python_profiler:
+            self.python_profiler.stop_profiling(
+                StepPhase.STEP_START,
+                end_mode=mode_keys_to_python_profile_mode(mode),
+                end_step=self.mode_steps[mode],
+            )
+            if self.profiler_config_parser.should_save_metrics(
+                MetricsCategory.PYTHON_PROFILING, self.mode_steps[mode]
+            ):
+                self.python_profiler.start_profiling(
+                    StepPhase.STEP_START,
+                    start_mode=mode_keys_to_python_profile_mode(mode),
+                    start_step=self.mode_steps[mode],
+                )
+
+    def _end_phase_python_profiling(self, mode):
+        if self.python_profiler:
+            self.python_profiler.stop_profiling(
+                StepPhase.STEP_END,
+                end_mode=mode_keys_to_python_profile_mode(mode),
+                end_step=self.mode_steps[mode],
+            )
+            if self.profiler_config_parser.should_save_metrics(
+                MetricsCategory.PYTHON_PROFILING, self.mode_steps[mode]
+            ):
+                self.python_profiler.start_profiling(
+                    StepPhase.STEP_END,
+                    start_mode=mode_keys_to_python_profile_mode(mode),
+                    start_step=self.mode_steps[mode],
+                )
+
+    def _begin_detailed_profiling(self, mode=ModeKeys.TRAIN):
```

Comment:
Same as above. I'd also recommend updating the function name (begin sounds misleading). Maybe `_handle_detailed_profiling`? And you can prepend all the helper functions with `_handle`.